app/src/main/java/org/sagebionetworks/research/mpower/ActivityRecorderService.kt

```diff
@@ -219,22 +236,47 @@ class ActivityRecorderService : DaggerService() {
     // Save to bridge - any recording is worth saving
     // Stop service
     private fun stopRecording() {
-        if (isRecording) {
+        if (state == RECORDING) {
             Log.d(
                 TAG,
-                "-- stopRecording ${SimpleDateFormat(TIME_PATTERN, Locale.US).format(Date())}"
+                "-- stopRecording ($state): ${SimpleDateFormat(TIME_PATTERN, Locale.US).format(Date())}"
             )
-            isRecording = false
-//            CoroutineScope(Dispatchers.Default).launch {
-//                recorderService.getActiveRecorders(taskUUID!!).forEach {
-//                    it.value.stop()
-//                }
-//            }
-            recorderManager.onStepTransition(stopStep, null, 0)
 
-            val sharedPrefs = getSharedPreferences(TRANSITION_PREFS, Context.MODE_PRIVATE)
-            sharedPrefs.edit().putLong(LAST_RECORDED_AT, Date().time).apply()
+            recorderManager?.onStepTransition(null, stopStep, NavDirection.SHIFT_RIGHT)
+            recordTimer.cancel()
+
+            state = RECORDED
+
+            if (elapsedTime >= MIN_RECORDING_TIME_SEC) {
```

Comment:
i haven't touched testing android for awhile now but there should be a way to instrument the android framework (since actual implementations are on device) and test this service.  i've used Robolectric a long time ago and still seems like the fastest way to run unit test on Android components without having to launch on device/emulator.

Reply:
We'll need to get into instrumented tests in order to verify the integration of the activity transition api with the RecorderManager but for this we should be able to write unit tests like the ones in app/src/test/java/org/sagebionetworks/research/mpower .